Sources/MongoSwift/BSON/AnyBSONValue.swift

```diff
@@ -2,8 +2,19 @@ import Foundation
 
 /// A struct wrapping a `BSONValue` type that allows for encoding/
 /// decoding `BSONValue`s of unknown type.  
-public struct AnyBSONValue: Codable, Equatable {
-    /// The `BSONValue` wrapped by this struct. 
+public struct AnyBSONValue: Codable, Equatable, Hashable {
+    // swiftlint:disable:next legacy_hashing
+    public var hashValue: Int {
+        if let doc = self.value as? Document {
+            return doc.extendedJSON.hashValue
+        } else {
+            let doc: Document = ["value": self.value]
+            // need to add some string to the beginning to ensure no collisions with the document case.
+            return ("EXT_JSON" + doc.extendedJSON).hashValue
```

Comment:
This seems like it would have a lot of overhead– why not use the naturally primitive hash values?

Reply:
It's the only way to guarantee that we minimize hash collisions. In some older versions of swift, using the primitive hash values led to some collisions. (see: https://travis-ci.org/mongodb/mongo-swift-driver/builds/478423525)

Whether or not this overhead is better than a potentially higher rate of hash collisions is unclear.

Reply:
I could be mistaken, but I thought the collisions were intentional so that naturally similar values were represented the same, e.g., `false` would have the same hash as `0`.

Reply:
I doubt hash collisions are ever intentional as they incur a performance hit, and hashmaps/dictionaries are generally unordered, so the clustering isn't particularly useful.

e.g. in 4.2 `false.hashValue` doesn't have the same hash as `0` (neither of them hash to 0 as they did in 4.0).

Reply:
I'm not sure managing that error (which is likely an artifact from objc days) warrants the overhead that will come along with having to convert to extended JSON to calculate a hash. Seems like most of the BSONValue classes/structs have simple hashes that wouldn't require calculation. I'm also not sure if there's a use case to use the hash of something like a bool.

Could we at least lazily convert to extendedJSON? Just seems like this is going to get expensive, fast.

Reply:
@jsflax are you positing that `("EXT_JSON" + doc.extendedJSON).hashValue` has significantly more overhead than `doc.extendedJSON.hashValue` ?

Reply:
The following is much faster (for non documents) and avoids most of the collision issues:
```
public var hashValue: Int {
        let prefix = String(UnicodeScalar(UInt8(self.value.bsonType.rawValue)))

        if let document = self.value as? Document {
            return "\(prefix)-\(document.extendedJSON)".hashValue
        } else {
            return "\(prefix)-\(self.value)".hashValue
        }
    }
```
I'm currently running into some collision issues specifically on Linux with Swift 4.0/4.1 that I'm trying to hammer out, but hopefully something along these lines will work. 

It's still converting the `BSONValue` to a string, but it is omitting the document creation and having to convert to extended JSON, for the most part.

Reply:
As it turns out, the Swift 4.0/4.1 `hashValue` for `String` on Linux seems to ignore non-ASCII bytes (e.g. the bson type I was prepending). I've opted for just prepending the ASCII representation of the bson type then.

This simple function minimizes collisions and is generally fast AFAIK (simple string conversion, no document creation). If performance is still a concern here, let me know @jsflax

Reply:
@mbroadst No, I'm positing that that converting to extended JSON would have a lot more overhead than, say, having an ObjectID and using the `.oid` hashValue. Or a string use it's own hashValue, and so on

Reply:
@patrickfreed That LGTM. Hope this didn't come off as an over optimization (tough to know without user feedback). Just seemed like there were simpler options

Reply:
Not at all, it was a concern I had too. The current method seems to be around ~2x as fast, so the change seems worthwhile.